modules/homa/src/lib.rs

```diff
@@ -16,118 +16,962 @@
+	#[pallet::getter(fn total_void_liquid)]
+	pub type TotalVoidLiquid<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// The total unclaimed redemption.
+	///
+	/// UnclaimedRedemption value: StakingCurrencyAmount
+	#[pallet::storage]
+	#[pallet::getter(fn unclaimed_redemption)]
+	pub type UnclaimedRedemption<T: Config> = StorageValue<_, Balance, ValueQuery>;
+
+	/// Requests to redeem staked currencies.
+	/// RedeemRequests: Map: AccountId => Option<(liquid_amount: Balance, addtional_fee: Rate)>
+	#[pallet::storage]
+	#[pallet::getter(fn redeem_requests)]
+	pub type RedeemRequests<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, (Balance, Rate), OptionQuery>;
+
+	/// The records of unbonding by AccountId.
+	///
+	/// Unbondings: double_map AccountId, ExpireEraIndex => UnboundingStakingCurrencyAmount
+	#[pallet::storage]
+	#[pallet::getter(fn unbondings)]
+	pub type Unbondings<T: Config> =
+		StorageDoubleMap<_, Twox64Concat, T::AccountId, Twox64Concat, EraIndex, Balance, ValueQuery>;
+
+	/// The weight for excution XCM msg on relaychain. Must be higher than
+	/// T::BaseXcmWeight + T::Weigher::weight(xcm_message_sended_by_homa)` on relaychain.
+	///
+	/// xcm_dest_weight: value: Weight
+	#[pallet::storage]
+	#[pallet::getter(fn xcm_dest_weight)]
+	pub type XcmDestWeight<T: Config> = StorageValue<_, Weight, ValueQuery>;
+
 	#[pallet::pallet]
 	pub struct Pallet<T>(_);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn integrity_test() {
+			assert!(!T::DefaultExchangeRate::get().is_zero());
+		}
+	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		/// Inject DOT to staking pool and mint LDOT in a certain exchange rate
-		/// decided by staking pool.
+		/// Mint liquid currency by put locking up amount of staking currency.
 		///
-		/// - `amount`: the DOT amount to inject into staking pool.
-		#[pallet::weight(<T as Config>::WeightInfo::mint())]
+		/// Parameters:
+		/// - `amount`: The amount of staking currency used to mint liquid currency.
+		#[pallet::weight(10_000)]
 		#[transactional]
 		pub fn mint(origin: OriginFor<T>, #[pallet::compact] amount: Balance) -> DispatchResult {
-			let who = ensure_signed(origin)?;
-			T::Homa::mint(&who, amount)?;
+			let minter = ensure_signed(origin)?;
+
+			// Ensure the amount is above the mint threshold.
+			ensure!(amount > T::MintThreshold::get(), Error::<T>::BelowMintThreshold);
+
+			// TODO: is that neccesary?
+			ensure!(
+				Self::get_total_staking_currency().saturating_add(amount) <= Self::get_staking_currency_soft_cap(),
+				Error::<T>::ExceededStakingCurrencySoftCap
+			);
+
+			T::Currency::transfer(T::StakingCurrencyId::get(), &minter, &Self::account_id(), amount)?;
+			ToBondPool::<T>::mutate(|pool| *pool = pool.saturating_add(amount));
+
+			// calculate the liquid amount by the current exchange rate.
+			let liquid_amount = Self::convert_staking_to_liquid(amount)?;
```

Comment:
this depends on `ToBondPool` right, so we should read it first before modify `ToBondPool`

should always read exchange rate before any actions to ensure the rate is not partially updated